shared/src/pool_fetching.rs

```diff
@@ -165,18 +166,49 @@ impl BaselineSolvable for Pool {
     }
 }
 
-// Read though Pool Fetcher that keeps previously fetched pools in a cache, which get invalidated whenever there is a new block
+const MAX_CACHED_BLOCKS: usize = 10;
+
+// Read though Pool Fetcher that keeps previously fetched pools in a LRU cache. Pools fetched for `BlockNumber::Latest` get invalidated whenever there is a new block
 pub struct CachedPoolFetcher {
     inner: Box<dyn PoolFetching>,
-    cache: Arc<Mutex<(Block, HashMap<TokenPair, Pool>)>>,
+    cache: Arc<Mutex<Cache>>,
     block_stream: CurrentBlockStream,
 }
 
+struct Cache {
+    /// Used to store details (e.g. hash) about the latest block. Needed so we know when we the Latest block change and
+    // we need to prune the cache entry for Block::Latest (which doesn't store any details)
+    latest_block: CurrentBlock,
+    pools: LruCache<Block, HashMap<TokenPair, Pool>>,
+}
+
+/// Helper type to work around the fact that ethcontract::BlockNumber is not Hashable
+#[derive(PartialEq, Eq, Hash)]
+enum Block {
+    Latest,
+    Pending,
```

Comment:
Maybe `Pending` shoulnd't be allowed in the cache. This way we always re-fetch for `pending` block?